### DISCUSSION TITLE
Notification Usage

### DISCUSSION THREAD

**Discussion by mightylastingcode** [Mair/esp32-course#100] [Course Qeustions] [answered]
Hi Mair,

Notification is introduced first and discussed in great details in your FreeRTOS course section.   It seems that it can emulate binary/counting semaphore, event flag groups, and queue as long as I am using it for directly communicate from a send task or an ISR routine to a receive task.   

1. a counting semaphore - I can use notification to count the occurrence of the event for the receiving task.
2. a binary semaphore - I can use notification to emulate give/take as in semaphore.
3. an event flag group - I can use notification to send multiple flag bits to the receiving task to do something.  
4. queue - I can send notification to send one 32 bit data or a character to another task.  The transfer is likely faster.

Am I correct?   Any downside from overusing it?

Should I always start with notification as the choice of direct communication between two tasks?  Such that I don't need to consider other communication objects like semaphore/event flag groups/queue...except for mutex.

Thanks,
Michael

**Comment by Mair** [comment, marked as answer]
Hi @mightylastingcode 
Yes correct. 
And to your point about notifications. According the the FreeRtos documentation, notifications are faster then regular semaphores although to be totally honest with you in 99% of my code I use 
* Queues for relaying data
* events for notification
* binary semaphores for simple notifications
* Mutex to guard resources. (these I tend to use a bit)
binary semaphores tend to turn into event notifications as the code gets more complex. For example. I want a notification when I connect to the wireless. This turns into I want an indication when I connect or get an error which turn into I want an indication when I connect, get an error or a time out.

**Reply by mightylastingcode** [reply]
Oh.   Yeah.  I went to the FreeRTOS website.  It looks like that some people in the Q&A forum are endorsing the use of notification because they said that the majority of the drivers only need direct communication between two tasks.    

I got so happy to discover something new and fast to use from your course.  I still am.    Like you, I use other communication objects in my last two projects.     I have not used binary semaphores because I prefer events for notification like you do.   I have multiple events occurred in the send task like your example for the wireless application.

I am going to explore the possible replacement with notification.  I got nothing to lose.   See if I can gain something from it.   It is still exciting because I did not know that it exists until now.    Just one more trick in my bag.   Hopefully it doesn't have any downside to worry about.

Thank you for your candid feedback!  Cheers!
Michael
